---
title: "HC2022"
author: "MBI"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("RSQLite")
library('RSQLite')
library(DBI)
library(tidyverse)

mydb <- dbConnect(RSQLite::SQLite(), "my-db.sqlite")
#dbDisconnect(mydb)
#con <- dbConnect(RSQLite::SQLite(), ":memory:")

#dbWriteTable(mydb, "mtcars", mtcars)
#dbListTables(mydb)

```



```{r project_code, include=FALSE}

# We need to pass the the project code to R using a file or as an option in the script. 
# This is important to make each MAGID unique as the output for multiple projects can be the same e.g MAG??

project_id = read.delim(file="Project_code.txt", header=FALSE)
project_id = as.character(project_id$V1)

```

```{r taxonomic_data, include=FALSE}

# This block parses the taxonomic summary based on GTDBTK


tax_table = read.delim2(file="metagenomics/taxonomic_annotations/gtdb-tk/gtdbtk.bac120.summary.tsv")
tax_table$project_id = project_id
rownames(tax_table) = tax_table$user_genome 
tax_table$mag_id = paste0(tax_table$project_id, "_", tax_table$user_genome) 
class_table = matrix(nrow = length(tax_table$user_genome), ncol=8) 
rownames(class_table) = tax_table$user_genome
colnames(class_table) = c("mag_id",
                          "domain", "phylum","class","order",
                          "family", "genus","species"
                          )
class_table = as.data.frame(class_table)
class_table$mag_id = tax_table$mag_id 


for (taxa in tax_table$user_genome) {
  classification_string = as.character(tax_table[taxa, "classification"])
  classification_string = strsplit(x=classification_string, split =  ";")
  
  # Get different parts of the string
  my_domain = gsub(classification_string[[1]][1], pattern = "^d__", replacement = "")
  my_phylum = gsub(classification_string[[1]][2], pattern = "^p__", replacement = "")
  my_class = gsub(classification_string[[1]][3], pattern = "^c__", replacement = "")
  my_order = gsub(classification_string[[1]][4], pattern = "^o__", replacement = "")
  my_family = gsub(classification_string[[1]][5], pattern = "^f__", replacement = "")
  my_genus = gsub(classification_string[[1]][6], pattern = "^g__", replacement = "")
  my_species = gsub(classification_string[[1]][7], pattern = "^s__", replacement = "")
  class_table[taxa, "domain"] = my_domain
  class_table[taxa, "phylum"] = my_phylum
  class_table[taxa, "class"] = my_class
  class_table[taxa, "family"] = my_family
  class_table[taxa, "order"] = my_order
  class_table[taxa, "genus"] = my_genus
  class_table[taxa, "species"] <- my_species
  }

tax_table2 = cbind(tax_table, class_table)
tax_table2
taxonomy = tax_table2[, c("mag_id", "user_genome",
                         "domain", "phylum","class","order",
                          "family", "genus","species")
                     ]

dbWriteTable(mydb, "taxonomy", taxonomy, overwrite=T)

```


```{r project_samples, include=FALSE}

# This block stores the names of the samples for a project 

samples = read.delim2(file="samples.tsv", header=T)
names(samples)[1] = "sample_id"
samples$project_id = project_id
samples_table = samples[, c("project_id", "sample_id")]
dbWriteTable(mydb, "project_sample", samples_table, overwrite=T)
```


```{r dram, include=FALSE}

# This block reads the anotations from DRAM 
# It also separates them into subtables just if needed

dram = read.delim2("functional_profiles/DRAM/annotations/annotations.tsv" )
names(dram)[1] = "GeneID"
dram$genome_id = paste0(project_id, "_", dram$fasta)
dram$fasta = dram$genome_id
dram$genome_id = NULL
names(dram)[2] = "genome_id"
dram$heme_regulatory_motif_count = NULL


dram_ko = dram %>%
  select(genome_id, GeneID, ko_id) %>%
  filter(ko_id != "")

dram_kegg = dram %>%
  select(genome_id, GeneID, kegg_hit) %>%
  filter(kegg_hit != "")

dram_pfam = dram %>%
  select(genome_id, GeneID, pfam_hits) %>%
  filter(pfam_hits != "")

dram_cazy = dram %>%
  select(genome_id, GeneID, cazy_id, cazy_hits) %>%
  filter(cazy_id != "") %>%
  filter(cazy_hits != "") %>%
  pivot_longer(cols=cazy_id:cazy_hits)



dram_peptidase = dram %>%
  select(genome_id, GeneID, peptidase_id, peptidase_family, peptidase_hit,
                            peptidase_RBH, peptidase_identity,
                            peptidase_bitScore, peptidase_eVal) %>%
  filter(peptidase_id != "") %>%
  pivot_longer(cols=peptidase_id:peptidase_eVal)
dram_peptidase

```

```{r bakta_annotations}

# Reads all the invidual MAG annotation tables, also separates the
# results from multiple databases into individual databases columns

baktas = list.files(path = "functional_profiles/", pattern = "MAG[0-9]+.tsv", full.names = T )
bakta_all = as.data.frame(matrix(ncol=10))
colnames(bakta_all)[10] = "MAGID"
for (bakta_file in baktas) {
temp_bakta = read.delim(file=bakta_file, header=F, comment.char = "#")
bakta_file = gsub(bakta_file, pattern = "[a-zA-Z0-9_/]+/", replacement="")
temp_bakta$MAGID = gsub(x=bakta_file, pattern=".tsv", replacement = "")
bakta_all = rbind(bakta_all, temp_bakta)
}


colnames(bakta_all) = c("Sequence_Id", "Type",
                        "start_position", "end_position", "Strand", 
                        "Locus_Tag", "Gene", "Product", "DbXrefs", 
                        "MAGID")

bakta = bakta_all[2:nrow(bakta_all), c("MAGID", "Sequence_Id", "Type",
                        "start_position", "end_position", "Strand", 
                        "Locus_Tag", "Gene", "Product", "DbXrefs")]

bakta$COG = NA
bakta$KEGG = NA
bakta$SO = NA
bakta$UniParc = NA
bakta$UniRef = NA
bakta$RefSeq = NA

counter= 1

while (counter <= nrow(bakta)){

  bakta_test = bakta$DbXrefs[counter]
  bakta_test2 =strsplit(x =  bakta_test, split = ", ")
  bakta_test3 = as.data.frame(bakta_test2)
  names(bakta_test3) = "Variable"
  bakta_test4 = separate(bakta_test3, col = Variable, sep=":", into =  c("DB", "Value"))
  
  COG_set = as.list(subset(bakta_test4, bakta_test4$DB == "COG")$Value)
  KEGG_set = as.list(subset(bakta_test4, bakta_test4$DB == "KEGG")$Value)
  SO_set = as.list(subset(bakta_test4, bakta_test4$DB == "SO")$Value)
  UniParc_set = as.list(subset(bakta_test4, bakta_test4$DB == "UniParc")$Value)
  UniRef_set = as.list(subset(bakta_test4, bakta_test4$DB == "UniRef")$Value)
  RefSeq_set = as.list(subset(bakta_test4, bakta_test4$DB == "RefSeq")$Value)
  
  if(length(COG_set) == 0){COG_set <-NA}
  if(length(KEGG_set) == 0){KEGG_set <-NA}
  if(length(SO_set) == 0){SO_set <-NA}
  if(length(UniParc_set) == 0){UniParc_set <-NA}
  if(length(UniRef_set) == 0){UniRef_set <-NA}
  if(length(RefSeq_set) == 0){RefSeq_set <-NA}
  
  #bakta_temp_results = data.frame(COG=toString(COG_set),
  #                                RefSeq=toString(RefSeq_set),
  #                                KEGG=toString(KEGG_set),
  #                                SO=toString(SO_set),
  #                                UniParc=toString(SO_set))
  #bakta_temp_results
  
  COG_set = toString(COG_set)
  KEGG_set = toString(KEGG_set)
  SO_set = toString(SO_set)
  UniParc_set = toString(UniParc_set)
  UniRef_set = toString(UniRef_set)
  RefSeq_set = toString(RefSeq_set)
  
  bakta$COG[counter] <- COG_set
  bakta$KEGG[counter] <- KEGG_set
  bakta$SO[counter] <- SO_set
  bakta$UniParc[counter] <- UniParc_set
  bakta$UniRef[counter] <- UniRef_set
  bakta$RefSeq[counter] <- RefSeq_set
  
  counter = counter + 1
  
}



```


```{r bakta_annotation_summary}

# Parses list of individual MAG annotation summaries from bakta


baktas2 = list.files(path = "functional_profiles/", pattern = "MAG[0-9]+.txt", full.names = T )

bakta_stats = as.data.frame(matrix(ncol=3))
names(bakta_stats) = c("Variable", "Value", "MAGID")

for (bakta_file in baktas2){

myfile = read_file(bakta_file)
myfile2 = strsplit(x =  myfile, split = "\n")
myfile2 = as.data.frame(myfile2)
names(myfile2)[1] = "Variable"
myfile2 = subset(myfile2, myfile2$Variable != "")
myfile3 = separate(myfile2, col = Variable, sep = ":", into=c("Variable", "Value"))
myfile3 = subset(myfile3, myfile3$Value != "")
MAG_name = gsub(bakta_file, pattern = "[a-zA-Z0-9_/]+/", replacement="")
MAG_name = gsub(MAG_name, pattern = ".txt$", replacement="")
myfile3$MAGID = paste0(MAG_name, "_" , project_id)
bakta_stats =rbind(bakta_stats, myfile3)
}

bakta_stats = bakta_stats [2:nrow(bakta_stats), c("MAGID", "Variable", "Value")]  
bakta_stats_wide = pivot_wider(data = bakta_stats, names_from = Variable, values_from = Value)
```


```{r join bakta and Dram}

# Joins the bakta and DRAM annotation tables based on the GeneID which is a combination 
# Of the MAGID and the Locus tag. 
# We may need to add the project  identifier to make it unique, though there is a column called genomeID which is # a combination of project and MAG



bakta$GeneID = paste0(bakta$MAGID,"_", bakta$Locus_Tag)
all_annotations = merge(bakta, dram, by="GeneID")

```