---
params:
   new_title: "My Title!"
title: "`r params$new_title`"
date: "`r Sys.Date()`"
output: 
  rmdformats::readthedown:
  toc: true
  toc_depth: 2
---

```{css, echo=FALSE}
.box1 {
  display: inline-block;
  width: 45%;
  overflow: auto;
  vertical-align: top;
}

.box1 img {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.figure {
  margin: auto;
  justify-content: center;
}
```


```{r message=FALSE, warning=FALSE, echo=FALSE}
#Formatting strain ID from strain name
strainid = as.numeric(gsub("MAG", "", strain))

#Database connection
con <- dbConnect(RSQLite::SQLite(), "StrainPassportReplica.db")

```

# 1. Taxonomy

<div class="box1">
```{r message=FALSE, warning=FALSE, echo=FALSE}
#Format Taxonomy Classification Information
Taxonomy = dbGetQuery(con, paste0("SELECT * FROM Taxonomy WHERE ID IN ( ", strainid ," )"))
classificationformat <- gsub(".__", "", Taxonomy$classification)
classification <- data.frame(str_split_fixed(classificationformat, ";", 7))
colnames(classification) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
SpeciesName = ifelse(classification$Species == "", "Unknown", as.character(classification$Species))
SpeciesName = gsub("_.", "", SpeciesName)

if(SpeciesName == "Glutamicibacter sp004320535") {
  SpeciesName = "Arthrobacter S41"
}

#Add Taxonomy Classification Table
t(classification) %>%
  kbl() %>%
  kable_paper("hover", full_width = F, position = "float_left", font_size = 16) %>%
  column_spec(1, bold = TRUE, width = "20em") %>%
  column_spec(2, italic = TRUE, width = "20em") %>%
  add_header_above(c("Taxonomy" = 2))
```
</div>

<div class="box1">
```{r, message=FALSE, warning=FALSE, echo=FALSE, out.width = "70%", fig.cap=paste0("Species: ", SpeciesName, " Source: commons.wikimedia.org")}
#Add Image (if it exists, if not add placeholder)
tryCatch(
    expr = {
       include_graphics(paste0("img/",strain, ".jpg"))
    },
    error = function(e){
       randimg=sample(c(1,2), size = 1)
       include_graphics(paste0("img/",randimg, ".png"))
    }
)
```
</div>

# 2. Assembly Statistics

<div class="box1" style="width: 35%;">
```{r message=FALSE, warning=FALSE, echo=FALSE}
AssemblyStats = dbGetQuery(con, paste0("SELECT * FROM AssemblyStats WHERE ID IN ( ", strainid ," )"))
AssemblyStats$GC <- paste0(round(as.numeric(AssemblyStats$gc_avg), 3), "Â±", round(as.numeric(AssemblyStats$gc_std), 3))
AssemblyStats <- AssemblyStats[c("n_contigs", "contig_bp", "ctg_N50", "ctg_L50", "ctg_max", "GC")]
colnames(AssemblyStats) <- c("N. Contigs", "Size (bp)", "N50 (bp)", "L50 (bp)", "Max Contig (bp)", "GC")

t(AssemblyStats) %>%
  kbl() %>%
  kable_paper("hover", full_width = F, position = "left") %>%
  column_spec(1, bold = TRUE, width = "10em") %>%
  column_spec(2, width = "10em") %>%
  add_header_above(c("Assembly Statistics" = 2))
```

```{r message=FALSE, warning=FALSE, echo=FALSE}
Completeness = dbGetQuery(con, paste0("SELECT * FROM Completeness WHERE ID IN ( ", strainid ," )"))
Completeness <- Completeness[c("Completeness", "Contamination", "Strain heterogeneity")]
t(Completeness) %>%
  kbl() %>%
  kable_paper("hover", full_width = F, position = "left") %>%
  column_spec(1, bold = TRUE, width = "10em") %>%
  column_spec(2, width = "10em") %>%
  add_header_above(c("Genome Completeness" = 2))
```
</div>

<div class="box1" style="width: 55%;">
```{r message=FALSE, warning=FALSE, echo=FALSE, fig.height = 6.2, fig.cap=paste(strain, " contig assembly")}
#Setup contig plots from gff
gff.strain <- gff[which(gff$Strain == strain), ]
gff.strain$contig <- gff.strain$seqnames
gff.strain$seqnames <- as.character(gff.strain$Strain)
gff.strain.start <- aggregate(gff.strain$start, by=list(gff.strain$contig), FUN = min)
gff.strain.end <- aggregate(gff.strain$end, by=list(gff.strain$contig), FUN = max)
gff.strain.contig <- cbind(gff.strain.start, gff.strain.end$x)

#Use this for the final image
# colnames(gff.strain.contig) <- c("seqnames", "start", "end")
# gff.strain.contig$strain <- strain
# gff.strain.contig$seqnames <- as.character(gff.strain.contig$seqnames)

#Use this chunk in development
colnames(gff.strain.contig) <- c("contig", "start", "end")
gff.strain.contig$seqnames <- strain
gff.strain.contig <- gff.strain.contig[c("seqnames", "start", "end", "contig")]


#Plot contigs
par(bg = "#fcfcfc", mar=c(0,0,0,0))
circos.clear()
circos.par(start.degree = 90, gap.degree=0)
circos.initializeWithIdeogram(gff.strain.contig, plotType = c("axis"))

#Coding Regions
circos.genomicTrackPlotRegion(gff.strain.contig, stack = FALSE, panel.fun = function(region, value, ...) {
  circos.genomicRect(region, value,
                     ytop = 1, ybottom = 0, #Plots ORF positions by strand 
                     col = "darkblue", border = "white")
}, track.height = 0.1, bg.border = "grey50", ylim = c(0,1))
```
</div>

# 3. Metabolism and Function

## KEGG Function Annotation

```{r message=FALSE, warning=FALSE, echo=FALSE, fig.cap=paste(strain, " KEGG Functional Classification")}
#Coalate KEGG hierarchy data
KO.level1.strain <- KO.level1[which(KO.level1$Strain == strain), ]
KO.level1.strain <- KO.level1.strain[which(KO.level1.strain$Catagory %!in% c("09150 Organismal Systems", "09140 Cellular Processes")), ]

KO.level2.strain <- KO.level2[which(KO.level2$Strain == strain), ]
KO.level2.strain <- KO.level2.strain[which(KO.level2.strain$Catagory %in% level2.catagories), ]


#Plot Level 1
KO.level1.plot <- ggplot(KO.level1.strain, aes(x = Catagory, y = N.genes)) +
  geom_bar(stat="identity", fill = "darkred", color = "black") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust =0.6, size = 6, color="black"),
        axis.text.y = element_text(size = 6, color="black"),
        plot.background = element_rect(fill = "#fcfcfc")) +
  ggtitle("KEGG Functional Hierarchies") +
        xlab(NULL) + 
  ylab("Genes Per Catagory")

#Plot Level 2
KO.level2.plot <- ggplot(KO.level2.strain, aes(x = Catagory, y = N.genes)) +
  geom_bar(stat="identity", fill = "darkred", color = "black") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust =0.6, size = 6, color="black"),
        axis.text.y = element_text(size = 6, color="black"),
        plot.background = element_rect(fill = "#fcfcfc")) +
    ggtitle("KEGG Functional Hierarchies") +
        xlab(NULL) + 
  ylab("Genes Per Catagory")


#Convert to plotly
KO.level1.plotly <- ggplotly(KO.level1.plot)
KO.level2.plotly <- ggplotly(KO.level2.plot)

subplot(KO.level1.plotly, KO.level2.plotly)
```

## Resistome

```{r message=FALSE, warning=FALSE, echo=FALSE}
strain.annotations <- annotations[which(annotations$fasta == strain), ]
ko01504.resistome.strain <- ko01504.resistome.DF[which(ko01504.resistome.DF$KO %in% strain.annotations$kegg_id), ]
colnames(ko01504.resistome.strain) <- c("name", "Type", "Level1", "Level2", "Description", "KO")

# 
# ko01504.resistome.strain <- ko01504.resistome.DF[which(ko01504.resistome.DF$name %in% strain.annotations$kegg_id), ]
# colnames(ko01504.resistome.strain) <- c("KO", "Type", "Level1", "Level2", "Description")

if(nrow(ko01504.resistome.strain) > 0) {
ko01504.resistome.strain %>%
  kbl() %>%
  kable_paper("hover", full_width = T, position = "left")
} else {
  cat("No Resistance Markers Detected")
}
```

## Toxins

```{r message=FALSE, warning=FALSE, echo=FALSE}
strain.annotations <- annotations[which(annotations$fasta == strain), ]
ko02042.toxins.strain <- ko02042.toxins.DF[which(ko02042.toxins.DF$KO %in% strain.annotations$kegg_id), ]
colnames(ko02042.toxins.strain) <- c("name", "Type", "Level1", "Level2", "Description", "KO")

if(nrow(ko02042.toxins.strain) > 0) {
ko02042.toxins.strain %>%
  kbl() %>%
  kable_paper("hover", full_width = T, position = "left")
} else {
  cat("No Toxin Markers Detected")
}
```

## Virulence

```{r message=FALSE, warning=FALSE, echo=FALSE}
virulence.annotations.strain <- virulence.annotations[which(virulence.annotations$fasta == strain), ]
virulence.annotations.strain <- virulence.annotations.strain[c("X", "kegg_id", "kegg_hit", "pfam_hits")]
KO.vir.strain <- KO.vir[which(KO.vir$KO %in% virulence.annotations.strain$kegg_id), ] 
colnames(virulence.annotations.strain) <- c("GeneID", "KO", "Description", "PFAM")

if(nrow(virulence.annotations.strain) > 0) {
virulence.annotations.strain %>%
  kbl() %>%
  kable_paper("hover", full_width = T, position = "left")
} else {
  cat("No Virulence Markers Detected")
}
```

## Carbon Metabolism

```{r message=FALSE, warning=FALSE, echo=FALSE}
pathview(gene.data = as.character(strain.annotations$kegg_id),
  pathway.id = "01200", 
  species = "ko",
  gene.idtype = "KEGG",
  kegg.native = TRUE,
  out.suffix = strain)

tryCatch(
    expr = {
       include_graphics(paste0("ko01200.", strain, ".png"))
    },
    error = function(e){
       cmap=F
    }
)
```

## Nitrogen Metabolism

```{r message=FALSE, warning=FALSE, echo=FALSE}
pathview(gene.data = as.character(strain.annotations$kegg_id),
  pathway.id = "00910", 
  species = "ko",
  gene.idtype = "KEGG",
  kegg.native = TRUE,
  out.suffix = strain)

tryCatch(
    expr = {
       include_graphics(paste0("ko00910.", strain, ".png"))
    },
    error = function(e){
       cmap=F
    }
)
```

# 4. Publications

```{r message=FALSE, warning=FALSE, echo=FALSE}
publications = FALSE
tryCatch(
    expr = {
       publications <- read.delim(paste0("pubmed/", gsub(" ", "_", SpeciesName), "_pubmed.txt"), sep = "\t", header = TRUE)
       pub.by.year <- aggregate(publications$Year, by=list(publications$Year), FUN=length)
       colnames(pub.by.year) <- c("Year", "Count")
       publications$LINK <- paste0("https://doi.org/", publications$DOI)
       
       #Plot Pubs by year
       pub.plot <- ggplot(pub.by.year, aes(x = Year, y = Count)) +
        geom_bar(stat="identity", fill = "deepskyblue2", color = "black") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust =0.6, size = 6, color="black"),
              axis.text.y = element_text(size = 6, color="black"),
              plot.background = element_rect(fill = "#fcfcfc")) +
          ggtitle("Publications by Year") +
              xlab(NULL)
       ggplotly(pub.plot)
      
    },
    error = function(e){
      publications = FALSE
       cat("No Publications Found in Source: Pubmed")
    }
)
```

```{r message=FALSE, warning=FALSE, echo=FALSE, results='asis'}
if(publications != FALSE) {
       #Publication Table
       pander::pandoc.header(paste0("Last 10 publications out of ", nrow(publications), " between 2000 and 2021"), level = 2)
}
```

```{r message=FALSE, warning=FALSE, echo=FALSE}
if(publications != FALSE) {
       head(publications[order(publications$Year, na.last = T, decreasing = T), c("FirstAuthor", "Year", "Title", "Journal", "Volume", "Pages", "LINK")], 10) %>%
       kbl() %>%
       kable_paper("hover", full_width = T, position = "left")
       }
```

```{r message=FALSE, warning=FALSE, echo=FALSE, results='asis'}
if(publications != F) {
pander::pandoc.p(paste0("Full publication list found in associated data file: **pubmed/", gsub(" ", "_", SpeciesName), "_pubmed.txt**"))
}
```

# 5. U.S. Patents

```{r message=FALSE, warning=FALSE, echo=FALSE}
patents = FALSE
tryCatch(
    expr = {
       patents <- read.delim(paste0("Patents/", gsub(" ", "_", SpeciesName), "_US_patents.txt"), sep = "\t", header = TRUE)
    },
    error = function(e){
      patents = FALSE
       cat("No US Patents Found")
    }
)
```

```{r message=FALSE, warning=FALSE, echo=FALSE, results='asis'}
if(patents != F) {
        x = nrow(patents)
       if(x > 10) x = 10
       pander::pandoc.header(paste0("First ", x, " patents out of ", nrow(patents)), level = 2)
}
```

```{r message=FALSE, warning=FALSE, echo=FALSE}
if(patents != F) {
       head(patents[ c("patent_number", "patent_title")], x) %>%
       kbl() %>%
       kable_paper("hover", full_width = T, position = "left")
}
```

```{r message=FALSE, warning=FALSE, echo=FALSE, results='asis'}
if(patents != F) {
pander::pandoc.p(paste0("Full patent list found in associated data file: **Patents/", gsub(" ", "_", SpeciesName), "_US_patents.txt**"))
}
```
